### PR TITLE
Build docker image for Code Runner Server

### LINKED ISSUE #7: Create image for code runner
- Use the generated Dockerfile to build a docker image for the code runner application.

### PR DESCRIPTION
**Pull request by balandi1**
- Initial code for building an image using the Dockerfile.

### CODE REVIEW

File: builder/builder.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package builder
+
+import (
+	"fmt"
+	"os/exec"
+)
+
+func BuildCodeRunnerImage() error {
+
+	cmdStr := "sudo docker build -t test_image:1.0 ."
```
Comment (pradykaushik) [inline builder/builder.go]: 1. You shouldn't need to use sudo. This makes me feel that we should add docker setup instruction links in the README with a couple of lines as docker is a pre-requisite.
2. Just having a command as a string is error prone and also very difficult to extend and add further features. Instead, we should build the different parts of the command separately. Would strongly recommend using the [docker client for golang](https://docs.docker.com/engine/api/sdk/).

File: configurations/configurations_test.go [hunk 2]
```diff
@@ -12,9 +12,9 @@ ENV GOFLAGS=-mod=vendor
 ENV GOMODULE=on
 COPY . /code-runner
 WORKDIR /code-runner
-RUN go build -o code-runner-server
+RUN git clone https://github.com/assignment-exec/code-runner.git && cd code-runner && make
 EXPOSE 8082
-CMD ["./code-runner-server"]
+CMD ["./code-runner/code-runner-server -port 8082"]
 
 `
```
Comment (pradykaushik) [inline configurations/configurations_test.go:19]: Just looking at the test code. It should also remove the docker file once testing is done.

**Review by pradykaushik — CHANGES_REQUESTED** [review]

File: builder/builder.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/docker/docker/client"
+)
+
+// Builds Docker Image for Code Runner.
+func BuildCodeRunnerImage() error {
```
Comment (pradykaushik) [inline builder/builder.go]: Possibly change this to `BuildImage()` as the image will eventually not just be for the code-runner.

File: builder/builder.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	writer := tar.NewWriter(tarBuffer)
+	defer func() {
+		err = writer.Close()
+		if err != nil {
+			log.Fatal(err)
+			return
+		}
+	}()
+
+	dockerFilepath := "Dockerfile"
```
Comment (pradykaushik) [inline builder/builder.go]: no magic strings!!
Maybe have it as a constant that is also used by the docker file creator. That way, it's consistent.

File: builder/builder.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	dockerFileReader, err := os.Open(dockerFilepath)
+	if err != nil {
+		log.Fatalf(" unable to open dockerfile: %v",err)
```
Comment (pradykaushik) [inline builder/builder.go]: extra space?

File: builder/builder.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		log.Fatalf("unable to write tar header: %v", err)
+	}
+	_, err = writer.Write(readDockerFile)
+	if err != nil {
+		log.Fatalf("unable to write tar body: %v", err)
+	}
+
+	// Use the tar as build context while building image.
+	dockerFileTarReader := bytes.NewReader(tarBuffer.Bytes())
```
Comment (pradykaushik) [inline builder/builder.go:60]: why are we using this tar as the build context? Also, what is referred to by 'tar' here?

File: builder/builder.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+	response, err := dockerClient.ImageBuild(
+		buildContext,
+		dockerFileTarReader,
+		types.ImageBuildOptions{
+			Dockerfile: dockerFilepath})
```
Comment (pradykaushik) [inline builder/builder.go]: should fit in one line.

**Review by pradykaushik — CHANGES_REQUESTED** [review]
Overall looks good.
Just minor nit picks and questions.

Also, please run `go fmt ./...` on the project. This should fix formatting of builder.go.

**Comment by pradykaushik** [comment]
Please also update the README.md accordingly.

File: configurations/configurations_test.go (hunk 2 above)
Comment (balandi1) [reply configurations/configurations_test.go:19]: Test code doesn't create a Dockerfile. The dockerfile instructions are written in a bytes array, which is then compared to the expected output.

File: builder/builder.go (hunk 3 above)
Comment (balandi1) [reply builder/builder.go]: Done

File: builder/builder.go (hunk 4 above)
Comment (balandi1) [reply builder/builder.go]: Done

File: builder/builder.go (hunk 5 above)
Comment (balandi1) [reply builder/builder.go]: Resolved

File: builder/builder.go (hunk 7 above)
Comment (balandi1) [reply builder/builder.go]: Done

File: builder/builder.go (hunk 6 above)
Comment (balandi1) [reply builder/builder.go:60]: Because the function expects a tar of the Dockerfile and other files(if any to be included) while building the image.

**Review by pradykaushik — APPROVED** [review]

**Pull request merged**